coderd/workspaceagentsrpc.go

```diff
@@ -0,0 +1,344 @@
+		AppHostname:                     api.AppHostname,
+		AgentInactiveDisconnectTimeout:  api.AgentInactiveDisconnectTimeout,
+		AgentFallbackTroubleshootingURL: api.DeploymentValues.AgentFallbackTroubleshootingURL.String(),
+		AgentStatsRefreshInterval:       api.AgentStatsRefreshInterval,
+		DisableDirectConnections:        api.DeploymentValues.DERP.Config.BlockDirect.Value(),
+		DerpForceWebSockets:             api.DeploymentValues.DERP.Config.ForceWebSockets.Value(),
+		DerpMapUpdateFrequency:          api.Options.DERPMapUpdateFrequency,
+		ExternalAuthConfigs:             api.ExternalAuthConfigs,
+
+		// Optional:
+		WorkspaceID:          build.WorkspaceID, // saves the extra lookup later
+		UpdateAgentMetricsFn: api.UpdateAgentMetrics,
+	})
+
+	closeCtx, closeCtxCancel := context.WithCancel(ctx)
+	go func() {
+		defer closeCtxCancel()
+		err := agentAPI.Serve(ctx, mux)
+		if err != nil {
+			api.Logger.Warn(ctx, "workspace agent RPC listen error", slog.Error(err))
+			_ = conn.Close(websocket.StatusInternalError, err.Error())
+			return
+		}
+	}()
+
+	pingFn(closeCtx, ensureLatestBuildFn)
+}
+
+func ensureLatestBuild(ctx context.Context, db database.Store, logger slog.Logger, rw http.ResponseWriter, workspaceAgent database.WorkspaceAgent) (func() error, database.WorkspaceBuild, bool) {
+	resource, err := db.GetWorkspaceResourceByID(ctx, workspaceAgent.ResourceID)
+	if err != nil {
+		httpapi.Write(ctx, rw, http.StatusBadRequest, codersdk.Response{
+			Message: "Internal error fetching workspace agent resource.",
+			Detail:  err.Error(),
+		})
+		return nil, database.WorkspaceBuild{}, false
+	}
+
+	build, err := db.GetWorkspaceBuildByJobID(ctx, resource.JobID)
+	if err != nil {
+		httpapi.Write(ctx, rw, http.StatusBadRequest, codersdk.Response{
+			Message: "Internal error fetching workspace build job.",
+			Detail:  err.Error(),
+		})
+		return nil, database.WorkspaceBuild{}, false
+	}
+
+	// Ensure the resource is still valid!
+	// We only accept agents for resources on the latest build.
+	ensureLatestBuild := func() error {
+		latestBuild, err := db.GetLatestWorkspaceBuildByWorkspaceID(ctx, build.WorkspaceID)
+		if err != nil {
+			return err
+		}
+		if build.ID != latestBuild.ID {
+			return xerrors.New("build is outdated")
+		}
+		return nil
+	}
+
+	err = ensureLatestBuild()
+	if err != nil {
+		logger.Debug(ctx, "agent tried to connect from non-latest build",
+			slog.F("resource", resource),
+			slog.F("agent", workspaceAgent),
+		)
+		httpapi.Write(ctx, rw, http.StatusForbidden, codersdk.Response{
+			Message: "Agent trying to connect from non-latest build.",
+			Detail:  err.Error(),
+		})
+		return nil, database.WorkspaceBuild{}, false
+	}
+
+	return ensureLatestBuild, build, true
+}
+
+func (api *API) agentConnectionUpdate(ctx context.Context, workspaceAgent database.WorkspaceAgent, workspaceID uuid.UUID, conn *websocket.Conn) (func(closeCtx context.Context, ensureLatestBuildFn func() error), bool) {
```

Comment:
This signature seems needlessly convoluted.  Why do we need to return a function to the caller, for the caller to just call it once?